s3torchconnector/src/s3torchconnector/dcp/fsdp_filesystem.py

```diff
@@ -0,0 +1,249 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  // SPDX-License-Identifier: BSD
+
+import io
+import logging
+import os
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Generator, Union, Optional
+
+from s3torchconnectorclient._mountpoint_s3_client import S3Exception
+from tenacity import (
+    retry,
+    stop_after_attempt,
+    wait_fixed,
+    wait_random,
+    retry_if_exception_type,
+    before_sleep_log,
+    after_log,
+)
+from torch.distributed.checkpoint.filesystem import (
+    FileSystemReader,
+    FileSystemWriter,
+    FileSystemBase,
+)
+
+from s3torchconnector._s3client import S3Client
+from s3torchconnector._s3dataset_common import parse_s3_uri
+
+logger = logging.getLogger(__name__)
+
+
+class S3FileSystem(FileSystemBase):
+    def __init__(self, region: str, s3_client: Optional[S3Client] = None) -> None:
+        self._path: Union[str, os.PathLike] = ""
+        self._client = s3_client if s3_client is not None else S3Client(region)
+
+    @contextmanager
+    def create_stream(
+        self, path: Union[str, os.PathLike], mode: str
+    ) -> Generator[io.IOBase, None, None]:
+        """
+        Create a stream for reading or writing to S3.
+
+        Args:
+            path (Union[str, os.PathLike]): The S3 path to read or write.
+            mode (str): The mode for the stream. Supports 'rb' for read mode and 'wb' for write mode.
+
+        Yields:
+            io.BufferedIOBase: A stream for reading or writing to S3.
+
+        Raises:
+            ValueError: If the mode is not 'rb' or 'wb'.
+        """
+        path = _path_to_str(path)
```

Comment:
Nit: Here and in other places, may be we can avoid reassigning to the same variable.